src/parser.cpp

```diff
@@ -4,11 +4,76 @@
 #include "unicode.cpp"
 #include "scheme.cpp"
 
-#include <cctype>
 #include <algorithm>
+#include <cctype>
+#include <string_view>
 
 namespace ada {
 
+  /**
+   * @see https://url.spec.whatwg.org/#concept-domain-to-ascii
+   */
+  std::optional<std::string_view> domain_to_ascii(const std::string_view input, bool be_strict = false) {
+    // TODO: Implement this
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-opaque-host-parser
+   */
+  std::optional<std::string_view> parse_opaque_host(std::string_view input) {
+    // TODO: Implement this
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-ipv6-parser
+   */
+  std::optional<std::string_view> parse_ipv6(std::string_view input) {
+    // TODO: Implement this
+  }
+
+  /**
+   * The host parser takes a scalar value string input
+   * with an optional boolean isNotSpecial (default false), and then runs these steps:
+   * @see https://url.spec.whatwg.org/#host-parsing
+   */
+  std::optional<std::string_view> parse_host(std::string_view input, bool is_not_special = false) {
+    // If input starts with U+005B ([), then:
+    if (strncmp(input.begin(), "[", 1) == 0) {
```

Comment:
strncmp seems overkill here. Any reason to think you can't do `input[0] == "["`?

Reply:
I didn't want to check the length of the input. I'll change it.

Reply:
The`strncmp` function does not know how long the strings are... it receives two pointers and a length parameter. It will always access `input[0]`. Always. Even if the input string is empty. 

In fact, we can look at the assembly output:

```
        cmp     BYTE PTR [rsi], 91
        sete    al
```

It is exactly equivalent to...

```
    input[0] == '[';
```

There is no difference whatsoever as far as the final result is concerned !

So if we think that input could be empty, we have to check. The strncmp function will not do it.

Now... if we know that input is representing a C string, then we know that it must be null terminated, in which case, it cannot be empty, it must at least have one character ('\0').

It is not always wise to assume that your strings must be null terminated. E.g., what if our input is a substring of a large input?